### ISSUE TITLE
missing data

### ISSUE THREAD

**Issue by jhx2002** [eclipse-iceoryx/iceoryx#581]
## Required information

**Operating system:**
E.g. Ubuntu 18.04 LTS

**Compiler version:**
E.g. GCC 7.4.0

One topic is published by one puber, and received by one subscriber.  
Repeatedly send 500 msgs every 100ms.
Missing last messages more than 10.

**Comment by elfenpiff** [comment]
From your description I guess the code looks somehow like the following example:
```cpp
// publisher
int main() {
  iox::runtime::PoshRuntime::initRuntime("publisher");
   iox::popo::Publisher<int> publisher({"some", "topic", "blubb"});
   
  while(true) {
     for(int i = 0; i < 500; ++i) {
       publisher.loan().and_then([&](auto &sample){ *sample = i; sample.publish(); };
     }
     std::this_thread::sleep_for(std::chrono::milliseconds(100));
  }
}

// subscriber
int main() {
    iox::runtime::PoshRuntime::initRuntime("subscriber");
    iox::popo::SubscriberOptions subscriberOptions;
    subscriberOptions.queueCapacity = 10U;
    iox::popo::Subscriber<int> subscriber({"some", "topic", "blubb"}, subscriberOptions);
   
    while(true) {
       subscriber.take().and_then([](auto & sample){ std::cout << "received " << *sample << std::endl; }
    }
}
```

If this is roughly your use case could you check that the `subscriberOptions.queueCapacity` is large enough. If this is not the case older samples are thrown away when the queue is full.
The current default maximum is 256 but this can maybe be insufficient for you. To increase the size you have to generate the cmake files again with:
```sh
cmake -Bbuild -Hiceoryx_meta -DIOX_MAX_CHUNKS_HELD_PER_SUBSCRIBER_SIMULTANEOUSLY=1024 
```

Then you can increase the queue capacity even further.

@jhx2002 did this work for you? If this does not solve your problem could you maybe adjust the example code so that it represents your use case better?

**Comment by elfenpiff** [comment]
There are even more options available if you encounter a similar problem in the future. At the moment they are not yet documented but you can see them here https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_posh/cmake/iceoryx_posh_deployment.cmake

**Comment by elBoberido** [comment]
We need to be able to detect if non-default parameters are used because it might influence the ABI. Currently we only check for the same version, git commit or build date. The build date could be used as compatibility check here but that is more restricted than necessary. In order to get reproducible builds we also would have to remove that option.

**Comment by ankitkk** [comment]
Thanks - I did run into a similar issue and was able to get around it by tweaking exactly what you mentioned.  On the subscriber side hasMissingData() triggers, which is nice. However I couldn't figure out a way to know if the publish is going to fail so the publisher can back off. 

I was expecting the else to hit in the following code, but it doesn't. hasMissingData on subscriber side does flip to true. 

`  auto result = publisher->loan(sizeof(MyType));
            if (!result.has_error())
            {
            }
            else
            {  // ??? }
            `
This is an amazing project which you all have opensourced and the use case goes way beyond autonomous driving / robotics.  Honestly, this can easily replace the whole Dbus debacle on linux desktops.

**Comment by budrus** [comment]
Thanks for your feedback @ankitkk. You are not the first one who asks for blocking the publisher, maybe we should think about an option to enable this. Running out of chunks or detecting a full queue on the subscriber side would be use cases to block the publisher and wait for subscribers to solve this issue. 
For automotive guys blocking the publisher when a subscribe isn't fast enough is normally no option. Therefore we didn't have this in mind. Letting the user decide would be a valid option

**Comment by MatthiasKillat** [comment]
Note: once the publisher has successfully loaned memory the publish itself should not fail (as long as RouDi is running). It will deliver the data to the internal subscriber queue.
If this queue is too small (which can be adjusted as shown via queueCapacity) the least recent samples will be discarded (LRU policy, like in a ring buffer) and cannot be taken from the subscriber.

This basically means that your publisher creates data faster than you consuming app is taking it from the subscriber). You will know this by calling hasMissingData on the subscriber, but there is no way to get those missed samples.

I do not think this can be handled differently, since it would require unbounded memory at some point.

Alternative semantics which are not implemented could only be on the publisher side not to deliver the data when the queue is full, returning it to caller. We decided against this since in many cases (image processing, point clouds, ...) working with stale-data is not useful, it is better to deliver the latest data reliably.

**Comment by elfenpiff** [comment]
@ankitkk 

Let's say we would implement the feature that the publisher informs you that the queue of the corresponding subscriber is full so that you can wait. How should it look like? One idea which came into my mind was something like:
```cpp
auto result = publisher->loan(sizeof(MyType));
if (!result.has_error())
{
   // would return a vector with a report for every subscribed subscriber where you can get the transmission 
   // report or some kind of subscriber state
   auto report = result.publishAndReport(); 
   for(auto &e : report) {
      //  on what kind of informations are you interested in? Would you only like to know if the subscribers 
      // queue is full or is there also something else?
      if ( e.isSubscriberQueueFull() ) { // check if the queue is full
          // what would you like to do here??
      }
   }
}
```

> Honestly, this can easily replace the whole Dbus debacle on linux desktops.

Btw, this is one of my dreams and I hope I will have the time to realize this in the near future. Any help here would be more then welcome!

**Comment by budrus** [comment]
@elfenpiff I wouldn't make things complicated. If people want to block the publisher thread if subscribers are not fast enough, we would just not return from `publish()` until the problem is solved. Sure  this can block forever but that's how it is.

**Comment by ankitkk** [comment]
@budrus  Yes, blocking publisher gated on something like shouldBlock bool, would be awesome.  

Although - more information about subscribers when they connect would be helpful - e.g queue size can give more hints to publisher on how to throttle themselves.  iox-roudi daemon knows about all this - maybe this information can also be disseminated on a "system" topic by the rox-roudi.

**Comment by budrus** [comment]
@ankitkk I created an issue for this as we came to the conclusion that this option should be supported #615

**Comment by mossmaurice** [comment]
Closing this issue as the blocking publisher was implemented for Almond v1.0. Feel free to reopen.
